### ISSUE TITLE
[Rule Request] Prefer flatMap instead of map { }.reduce([], +)

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#2883]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

`flatMap` can be used instead of `map { ... }.reduce([], +)` to avoid iterating the collection twice.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```swift
let foo = bar.map { $0.array }.reduce([], +) // would trigger

let foo = bar.map { $0.int }.reduce(0, +) // ok
let foo = bar.flatMap { $0.array } // ok
``` 

3. Should the rule be configurable, if so what parameters should be configurable?

Just the severity.

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Opt-in as this can have false positives
